### PR TITLE
Clessig/develop/fix kcrps 1077

### LINKED ISSUE #1077: kernel CRPS implementation is outdated
### What happened?

`kernel_crps` loss function is outdated and no longer compatible with current code.

### What are the steps to reproduce the bug?

User kernel_crps as loss fct

### Hedgedoc link to logs and more information. This ticket is public, do not attach files directly.

_No response_

### PR DESCRIPTION
**Pull request by clessig**
## Description

Fix kernel CRPS loss function

## Issue Number

Closes #1077 

Is this PR a draft? Mark it as draft.

## Checklist before asking for review

-   [X] I have performed a self-review of my code
-   [X] My changes comply with basic sanity checks:
      - I have fixed formatting issues with `./scripts/actions.sh lint`
      - I have run unit tests with `./scripts/actions.sh unit-test`
      - I have documented my code and I have updated the docstrings.
      - I have added unit tests, if relevant
-   [X] I have tried my changes with data and code:
      - I have run the integration tests with `./scripts/actions.sh integration-test`
      - (bigger changes) I have run a full training and I have written in the comment the run_id(s): `launch-slurm.py --time 60`
      - (bigger changes and experiments) I have shared a hegdedoc in the github issue with all the configurations and runs for this experiments
-   [ ] I have informed and aligned with people impacted by my change:
    - for config changes: the MatterMost channels and/or a design doc
    - for changes of dependencies: the MatterMost software development channel

### CODE REVIEW

File: src/weathergen/train/loss.py [hunk 1]
```diff
@@ -70,18 +70,48 @@ def mse_ens(target, ens, mu, stddev):
...
+    Params:
+    target : shape ( num_data_points , num_channels )
+    pred : shape ( ens_dim , num_data_points , num_channels)
+    weights_channels : shape = (num_channels,)
+    weights_points : shape = (num_data_points)
+
+    """
+
+    preds = pred.permute([2, 1, 0]).unsqueeze(0).to(torch.float32)
+    targets = target.permute([1, 0]).unsqueeze(0).to(torch.float32)
```
Comment (shmh40) [inline src/weathergen/train/loss.py]: Why do we this permutation? And we unsqueeze for a possible future batch/something else (?) dimension?

File: src/weathergen/train/loss.py [hunk 2]
```diff
@@ -70,18 +70,48 @@ def mse_ens(target, ens, mu, stddev):
...
+
+    ens_size = preds.shape[-1]
+    mae = torch.mean(torch.abs(targets[..., None] - preds), dim=-1)
 
     if ens_size == 1:
         return mae
```
Comment (shmh40) [inline src/weathergen/train/loss.py]: Should we return mse, and mse_chs here to keep it consistent?

File: src/weathergen/train/loss.py [hunk 3]
```diff
@@ -70,18 +70,48 @@ def mse_ens(target, ens, mu, stddev):
...
+    ens_var = torch.zeros(size=preds.shape[:-1], device=preds.device)
+    # loop to reduce memory usage
+    for i in range(ens_size):
+        ens_var += torch.sum(ens_n * abs(preds[..., i].unsqueeze(-1) - preds[..., i + 1 :]), dim=-1)
+
+    # apply weighting
+    kcrps_locs_chs = mae + ens_var
+    if weights_points is not None:
+        kcrps_locs_chs = kcrps_locs_chs * weights_points
+    kcrps_chs = torch.mean(torch.mean(kcrps_locs_chs, 0), -1) * weights_channels
```
Comment (shmh40) [inline src/weathergen/train/loss.py]: Need an if statement here for the weights_channels. 

if weight_channels is not None: ...

File: src/weathergen/train/loss.py [hunk 4]
```diff
@@ -70,18 +70,48 @@ def mse_ens(target, ens, mu, stddev):
...
+    Compute kernel CRPS
+
+    Params:
+    target : shape ( num_data_points , num_channels )
+    pred : shape ( ens_dim , num_data_points , num_channels)
+    weights_channels : shape = (num_channels,)
+    weights_points : shape = (num_data_points)
+
+    """
+
```
Comment (shmh40) [inline src/weathergen/train/loss.py:93]: Mask out nans as in mse_channel_location_weighted?

File: src/weathergen/train/loss.py [hunk 5]
```diff
@@ -70,18 +70,48 @@ def mse_ens(target, ens, mu, stddev):
...
+):
+    """
+    Compute kernel CRPS
+
+    Params:
+    target : shape ( num_data_points , num_channels )
+    pred : shape ( ens_dim , num_data_points , num_channels)
+    weights_channels : shape = (num_channels,)
+    weights_points : shape = (num_data_points)
+
```
Comment (shmh40) [inline src/weathergen/train/loss.py:88]: Returns:
        loss: scalar - overall weighted CRPS
        loss_chs: [C] - per-channel CRPS (location-weighted, not channel-weighted)
    """

File: src/weathergen/train/loss.py (hunk 3 above)
Comment (shmh40) [reply src/weathergen/train/loss.py]: Also, not quite sure on the order of operations here -- @Jubeku 
is the order of aggregration and multiplication per channel in line with what we expect?

File: src/weathergen/train/loss.py (hunk 2 above)
Comment (Jubeku) [reply src/weathergen/train/loss.py]: Agreed, otherwise we get an error in the loss calculator if we only return a single value: 
`loss, loss_chs = loss_fct(...)`

File: src/weathergen/train/loss_calculator.py [hunk 6]
```diff
@@ -160,7 +160,7 @@ def _loss_per_loss_function(
 
             # accumulate loss
             loss_lfct = loss_lfct + loss
-            losses_chs += loss_chs.detach()
+            losses_chs = losses_chs + loss_chs.detach() if len(loss_chs) > 0 else losses_chs
```
Comment (Jubeku) [inline src/weathergen/train/loss_calculator.py:163]: How are you going to use the ensemble `loss_chs`? At the moment it is basically ignored.

File: src/weathergen/train/loss.py (hunk 3 above)
Comment (Jubeku) [reply src/weathergen/train/loss.py]: > Need an if statement here for the weights_channels.
> 
> if weight_channels is not None: ...

Actually, `weights_channels` default to 1.0 so it should be fine without the if statement.

File: src/weathergen/train/loss.py [hunk 7]
```diff
@@ -70,18 +70,48 @@ def mse_ens(target, ens, mu, stddev):
...
+    """
+
+    preds = pred.permute([2, 1, 0]).unsqueeze(0).to(torch.float32)
+    targets = target.permute([1, 0]).unsqueeze(0).to(torch.float32)
+
+    ens_size = preds.shape[-1]
+    mae = torch.mean(torch.abs(targets[..., None] - preds), dim=-1)
 
     if ens_size == 1:
```
Comment (Jubeku) [inline src/weathergen/train/loss.py]: Do we need to apply channel and location weights here in case of `ens_size==1`?

File: src/weathergen/train/loss.py (hunk 5 above)
Comment (Jubeku) [reply src/weathergen/train/loss.py:88]: > ```
> Returns:
>     loss: scalar - overall weighted CRPS
>     loss_chs: [C] - per-channel CRPS (location-weighted, not channel-weighted)
> """
> ```

@shmh40, loss_chs is also channel-weighted right? Are you sure it is the per-channel CRPS or rather the per-ensemble member CRPS loss?

File: src/weathergen/train/loss.py (hunk 1 above)
Comment (clessig) [reply src/weathergen/train/loss.py]: Simplifies computations below. In the MSE code we also need similar ones. But I added a comment.

File: src/weathergen/train/loss_calculator.py (hunk 6 above)
Comment (clessig) [reply src/weathergen/train/loss_calculator.py:163]: It's used for logging.
